py/objstr.c

```diff
@@ -1950,6 +1950,93 @@ STATIC mp_obj_t str_encode(size_t n_args, const mp_obj_t *args) {
 #endif
 
+#if MICROPY_PY_BUILTINS_BYTES_HEX
+STATIC mp_obj_t bytes_hex(size_t n_args, const mp_obj_t *args, const mp_obj_type_t *type) {
+    // First argument is the data to convert.
+    // Second argument is an optional separator to be used between values.
+    const char *sep = NULL;
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(args[0], &bufinfo, MP_BUFFER_READ);
+
+    // Code below assumes non-zero buffer length when computing size with
+    // separator, so handle the zero-length case here.
+    if (bufinfo.len == 0) {
+        return mp_const_empty_bytes;
+    }
+
+    vstr_t vstr;
+    size_t out_len = bufinfo.len * 2;
+    if (n_args > 1) {
+        // 1-char separator between hex numbers
+        out_len += bufinfo.len - 1;
+        sep = mp_obj_str_get_str(args[1]);
+    }
+    vstr_init_len(&vstr, out_len);
+    byte *in = bufinfo.buf, *out = (byte *)vstr.buf;
+    for (mp_uint_t i = bufinfo.len; i--;) {
+        byte d = (*in >> 4);
+        if (d > 9) {
+            d += 'a' - '9' - 1;
+        }
+        *out++ = d + '0';
+        d = (*in++ & 0xf);
+        if (d > 9) {
+            d += 'a' - '9' - 1;
+        }
+        *out++ = d + '0';
+        if (sep != NULL && i != 0) {
+            *out++ = *sep;
+        }
+    }
+    return mp_obj_new_str_from_vstr(type, &vstr);
+}
+
+// For binascii.hexlify.
+STATIC mp_obj_t bytes_hex_as_bytes(size_t n_args, const mp_obj_t *args) {
+    return bytes_hex(n_args, args, &mp_type_bytes);
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mp_obj_bytes_hex_as_bytes_obj, 1, 2, bytes_hex_as_bytes);
+
+// For bytes.hex and memoryview.hex.
+STATIC mp_obj_t bytes_hex_as_str(size_t n_args, const mp_obj_t *args) {
+    return bytes_hex(n_args, args, &mp_type_str);
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mp_obj_bytes_hex_as_str_obj, 1, 2, bytes_hex_as_str);
+
+STATIC mp_obj_t bytes_fromhex(mp_obj_t data) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+
+    if ((bufinfo.len & 1) != 0) {
+        mp_raise_ValueError(MP_ERROR_TEXT("odd-length string"));
+    }
+    vstr_t vstr;
+    vstr_init_len(&vstr, bufinfo.len / 2);
+    byte *in = bufinfo.buf, *out = (byte *)vstr.buf;
+    byte hex_byte = 0;
+    for (mp_uint_t i = bufinfo.len; i--;) {
+        byte hex_ch = *in++;
+        if (unichar_isxdigit(hex_ch)) {
+            hex_byte += unichar_xdigit_value(hex_ch);
+        } else {
+            mp_raise_ValueError(MP_ERROR_TEXT("non-hex digit found"));
+        }
+        if (i & 1) {
+            hex_byte <<= 4;
+        } else {
+            *out++ = hex_byte;
+            hex_byte = 0;
+        }
+    }
+    return mp_obj_new_str_from_vstr(&mp_type_bytes, &vstr);
+}
+// For binascii.unhexlify.
+MP_DEFINE_CONST_FUN_OBJ_1(mp_obj_bytes_fromhex_obj, bytes_fromhex);
+// For bytes.fromhex.
+STATIC MP_DEFINE_CONST_STATICMETHOD_OBJ(bytes_fromhex_staticmethod_obj, MP_ROM_PTR(&mp_obj_bytes_fromhex_obj));
```

Comment:
is it possible to do away with this staticmethod wrapper and for it still to work, because it's accessing the function via the class (not an instance)?  ie `bytes.fromhex` vs `b''.fromhex`?

Reply:
To answer my own question, yes it is.  So `bytes.fromhex` would work but `b''.fromhex` would not.

But actually in CPython this is a classmethod, and using classmethod means that the first argument will be the type, `bytes` vs `bytearray`.  Then you could make `bytearray.fromhex` work!

Reply:
Awesome. Slightly complicated by keeping the version that works for binascii, but overall this worked nicely. In the end it's about 24 extra bytes for the wrapper functions etc, so that brings the PR to +104 !

Now that it's a bit clearer which wrappers and the function objects are for binascii vs str/bytes/bytearray/memoryview, I've moved the fun obj macros into binascii, and so it's just the underlying functions `bytes_hex` and `bytes_fromhex` that are "public". Should they be renamed to `mp_obj_bytes_hex/fromhex`?

Reply:
Yes please.  If a function is not `STATIC` then it should be prefixed with `mp_` in some way.

Reply:
Done.